grafana/resource_library_panel.go

```diff
@@ -0,0 +1,260 @@
+				Summary:  fmt.Sprintf("Library Panel %q is in state, but no longer exists in grafana", panel.Name),
+				Detail:   fmt.Sprintf("%q will be recreated when you apply", panel.Name),
+			})
+			d.SetId("")
+			return diags
+		} else {
+			return diag.FromErr(err)
+		}
+	}
+
+	modelJSONBytes, err := json.Marshal(panel.Model)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	remotePanelJSON, err := unmarshalLibraryPanelModelJSON(string(modelJSONBytes))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	modelJSON := normalizeLibraryPanelModelJSON(remotePanelJSON)
+
+	d.SetId(panel.UID)
+	d.Set("uid", panel.UID)
+	d.Set("panel_id", panel.ID)
+	d.Set("org_id", panel.OrgID)
+	d.Set("folder_id", panel.Folder)
+	d.Set("description", panel.Description)
+	d.Set("type", panel.Type)
+	d.Set("name", panel.Name)
+	d.Set("model_json", modelJSON)
+	d.Set("version", panel.Version)
+	d.Set("folder_name", panel.Meta.FolderName)
+	d.Set("folder_uid", panel.Meta.FolderUID)
+	d.Set("created", panel.Meta.Created.String())
+	d.Set("updated", panel.Meta.Updated.String())
+
+	return diags
+}
+
+func UpdateLibraryPanel(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client).gapi
+	uid := d.Id()
+	panel := makeLibraryPanel(d)
+
+	resp, err := client.PatchLibraryPanel(uid, panel)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(resp.UID)
+	d.Set("uid", resp.UID)
+	return ReadLibraryPanel(ctx, d, meta)
+}
+
+func DeleteLibraryPanel(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client).gapi
+	uid := d.Id()
+	_, err := client.DeleteLibraryPanel(uid)
+	var diags diag.Diagnostics
+	if err != nil && !strings.HasPrefix(err.Error(), "status: 404") {
+		return diag.FromErr(err)
+	}
+	return diags
+}
+
+func makeLibraryPanel(d *schema.ResourceData) gapi.LibraryPanel {
+	modelJSON := d.Get("model_json").(string)
+	panelJSON, err := unmarshalLibraryPanelModelJSON(modelJSON)
+
+	panel := gapi.LibraryPanel{
+		UID:    d.Get("uid").(string),
+		Name:   d.Get("name").(string),
+		Folder: int64(d.Get("folder_id").(int)),
+		Model:  panelJSON,
+	}
+	if err != nil {
+		return panel
+	}
+	return panel
+}
+
+// unmarshalLibraryPanelModelJSON is a convenience func for unmarshalling
+// `model_json` field.
+func unmarshalLibraryPanelModelJSON(modelJSON string) (map[string]interface{}, error) {
+	unmarshalledJSON := map[string]interface{}{}
+	err := json.Unmarshal([]byte(modelJSON), &unmarshalledJSON)
+	if err != nil {
+		return nil, err
+	}
+	return unmarshalledJSON, nil
+}
+
+// validateLibraryPanelModelJSON is the ValidateFunc for `model_json`. It
+// ensures its value is valid JSON.
+func validateLibraryPanelModelJSON(model interface{}, k string) ([]string, []error) {
+	modelJSON := model.(string)
+	modelMap := map[string]interface{}{}
+	err := json.Unmarshal([]byte(modelJSON), &modelMap)
+	if err != nil {
+		return nil, []error{err}
+	}
+	return nil, nil
```

Comment:
You can reuse `unmarshalLibraryPanelModelJSON` here
```suggestion
	if _, err := unmarshalLibraryPanelModelJSON(modelJSON); err != nil {
		return nil, []error{err}
	}
	return nil, nil
```